frame/transaction-storage/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+		MissingStateData,
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			let transaction_count = if obsolete > Zero::zero() {
+				<TransactionCount<T>>::get(obsolete)
+			} else {
+				0
+			};
+			// return weight for `on_finalize` does
+			T::WeightInfo::on_finalize(transaction_count)
+		}
+
+		fn on_finalize(n: T::BlockNumber) {
+			<TransactionCount<T>>::insert(n, <Counter<T>>::get());
+			<Counter<T>>::kill();
+			// Drop obsolete roots
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			if obsolete > Zero::zero() {
+				<TransactionRoots<T>>::remove_prefix(obsolete);
+				<TransactionCount<T>>::remove(obsolete);
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Index and store data on chain. Minimum data size is 1 bytes, maximum is `MAX_DATA_SIZE`.
+		/// Data will be removed after `STORAGE_PERIOD` blocks, unless `renew` is called.
+		/// # <weight>
+		/// - n*log(n) of data size, as all data is pushed to an in-memory trie.
+		/// Additionally contains a DB write.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::store(data.len() as u32))]
+		pub(super) fn store(
+			origin: OriginFor<T>,
+			data: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			ensure!(data.len() > 0, Error::<T>::EmptyTransaction);
+			Self::apply_fee(origin, data.len() as u32)?;
+
+			// Chunk data and compute storage root
+			let chunks = data.chunks(CHUNK_SIZE).map(|c| c.to_vec()).collect();
+			let root = sp_io::trie::blake2_256_ordered_root(chunks);
+
+			let content_hash = sp_io::hashing::blake2_256(&data);
+			let extrinsic_index = <frame_system::Pallet<T>>::extrinsic_index().unwrap();
+			sp_io::transaction_index::index(extrinsic_index, data.len() as u32, content_hash);
+
+			let block_num = <frame_system::Pallet<T>>::block_number();
+			let counter = <Counter<T>>::mutate(|c| { *c += 1; *c - 1 });
+			<TransactionRoots<T>>::insert(block_num, counter, TransactionInfo {
+				chunk_root: root,
+				size: data.len() as u32,
+				content_hash: content_hash.into(),
+			});
+			Self::deposit_event(Event::Stored(counter));
+			Ok(().into())
+		}
+
+		/// Renew previously stored data. Parameters are the block number that contains
+		/// previous `store` or `renew` call and transaction index within that block.
+		/// Transaction index is emitted in the `Stored` or `Renewed` event.
+		/// Applies same fees as `store`.
+		/// # <weight>
+		/// - Constant.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::renew())]
+		pub(super) fn renew(
+			origin: OriginFor<T>,
+			block: T::BlockNumber,
+			index: u32,
+		) -> DispatchResultWithPostInfo {
+			let info = <TransactionRoots<T>>::get(block, index).ok_or(Error::<T>::RenewedNotFound)?;
```

Comment:
would be nice to move the origin check the top of this function, and then potentially we can fail early without needing to do a storage read.

Reply:
I would say in general, it is a good idea just to move the origin check into each of the extrinsics and then pass sender to `apply_fee`